pkg/auth/providers/genericoidc/genericoidc_provider.go

```diff
@@ -0,0 +1,165 @@
+		DisplayName:   searchValue,
+		LoginName:     searchValue,
+		PrincipalType: principalType,
+		Provider:      g.Name,
+	}
+
+	principals = append(principals, p)
+	return principals, nil
+}
+
+func (g *GenOIDCProvider) GetPrincipal(principalID string, token v3.Token) (v3.Principal, error) {
+	var p v3.Principal
+
+	// parsing id to get the external id and type. Example genericoidc_<user|group>://<user sub | group name>
+	var externalID string
+	parts := strings.SplitN(principalID, ":", 2)
+	if len(parts) != 2 {
+		return p, errors.Errorf("invalid id %v", principalID)
+	}
+	externalID = strings.TrimPrefix(parts[1], "//")
+	parts = strings.SplitN(parts[0], "_", 2)
+	if len(parts) != 2 {
+		return p, errors.Errorf("invalid id %v", principalID)
+	}
+	provider := parts[0]
+	principalType := parts[1]
+	if externalID == "" && principalType == "" {
+		return p, fmt.Errorf("invalid id %v", principalID)
+	}
+	if principalType != UserType && principalType != GroupType {
+		return p, fmt.Errorf("invalid principal type: %s", principalType)
+	}
+	if principalType == UserType {
+		p = v3.Principal{
+			ObjectMeta:    metav1.ObjectMeta{Name: provider + "_" + principalType + "://" + externalID},
+			DisplayName:   externalID,
+			LoginName:     externalID,
+			PrincipalType: UserType,
+			Provider:      g.Name,
+		}
+	} else {
+		p = g.groupToPrincipal(externalID)
+	}
+	p = g.toPrincipalFromToken(principalType, p, &token)
+	return p, nil
+}
+
+// TransformToAuthProvider yields information used, typically by the UI, to be able to form URLs used to perform login.
+func (g *GenOIDCProvider) TransformToAuthProvider(authConfig map[string]interface{}) (map[string]interface{}, error) {
+	p := common.TransformToAuthProvider(authConfig)
+	p[publicclient.GenericOIDCProviderFieldRedirectURL] = g.getRedirectURL(authConfig)
+	p[publicclient.GenericOIDCProviderFieldScopes] = authConfig["scope"]
+	return p, nil
+}
+
+// RefetchGroupPrincipals is not implemented for OIDC.
+func (g *GenOIDCProvider) RefetchGroupPrincipals(principalID string, secret string) ([]v3.Principal, error) {
+	return nil, errors.New("Not implemented")
+}
+
+// groupToPrincipal takes a bare group name and turns it into a v3.Principal group object by filling-in other fields
+// with basic provider information.
+func (g *GenOIDCProvider) groupToPrincipal(groupName string) v3.Principal {
+	p := v3.Principal{
+		ObjectMeta:    metav1.ObjectMeta{Name: g.Name + "_" + GroupType + "://" + groupName},
+		DisplayName:   groupName,
+		Provider:      g.Name,
+		PrincipalType: GroupType,
+		Me:            false,
+	}
+	return p
+}
+
+// getRedirectURL uses the AuthConfig map to build-up the redirect URL passed to the OIDC provider at login-time.
+func (g *GenOIDCProvider) getRedirectURL(config map[string]interface{}) string {
+	authURL, _ := baseoidc.FetchAuthURL(config)
+
+	return fmt.Sprintf(
+		"%s?client_id=%s&response_type=code&redirect_uri=%s",
+		authURL,
+		config["clientId"],
+		config["rancherUrl"],
```

Comment:
These need to be query encoded, or use `AuthCodeURL`, as Kevin suggests, which already does it.